File: solutions/csharp/binary-search-tree/1/BinarySearchTree.cs
using System.Collections;

public class BinarySearchTree : IEnumerable<int>
{
    public BinarySearchTree(int value) => Value = value;

    public BinarySearchTree(IEnumerable<int> values)
    {
        using var enumerator = values.GetEnumerator();
        enumerator.MoveNext();

        Value = enumerator.Current;

        while (enumerator.MoveNext())
            Add(enumerator.Current);
    }

    public int Value { get; }

    public BinarySearchTree? Left { get; private set; }

    public BinarySearchTree? Right { get; private set; }

    public BinarySearchTree Add(int value)
    {
        if (value <= Value)
            Left = Left?.Add(value) ?? new BinarySearchTree(value);
        else
            Right = Right?.Add(value) ?? new BinarySearchTree(value);

        return this;
    }

    public IEnumerator<int> GetEnumerator()
    {
        if (Left is not null)
        {
            foreach (var value in Left)
                yield return value;
        }

        yield return Value;

        if (Right is not null)
        {
            foreach (var value in Right)
                yield return value;
        }
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}

File: solutions/fsharp/binary-search-tree/1/BinarySearchTree.fs
module BinarySearchTree

type Tree =
    {
        Data: int
        Left: Tree option
        Right: Tree option
    }

let left node = node.Left

let right node = node.Right

let data node = node.Data

let private leaf value =
    Some { Data = value; Left = None; Right = None }

let private insertChild insertFn value branch =
    branch |> Option.map (insertFn value) |> Option.orElse (leaf value)

let rec private insert value node =
    if value <= node.Data then
        { node with Left = insertChild insert value node.Left }
    else
        { node with Right = insertChild insert value node.Right }

let create items =
    match items with
    | [] -> invalidArg "items" "A tree requires at least one value."
    | root :: values ->
        let initial = { Data = root; Left = None; Right = None }
        values |> List.fold (fun tree value -> insert value tree) initial

let sortedData node =
    let rec traverse current result =
        match current with
        | None -> result
        | Some currentNode ->
            let result = traverse currentNode.Right result
            traverse currentNode.Left (currentNode.Data :: result)

    traverse (Some node) []
